### PR TITLE
Adds support for `componentDidCatch()`

### PR DESCRIPTION
**Pull request by johnhaitas**
This does not include support for the `info` argument because `preact`'s support for `componentDidCatch()` does not pass this argument (https://github.com/developit/preact/pull/819#issue-136629967)

Also includes 3 new tests.

### CODE REVIEW

**Comment by johnhaitas** [comment]
This should resolve https://github.com/developit/preact-render-to-string/issues/64

**Comment by johnhaitas** [comment]
This current implementation is incorrect per this documentation: https://reactjs.org/docs/react-component.html#error-boundaries
> **NOTE**
> Error boundaries only catch errors in the components *below* them in the tree. An error boundary can’t catch an error within itself.

Updating the code, starting with the tests.

**Comment by johnhaitas** [comment]
I have updated the tests to verify that the error is in fact being passed to `componentDidCatch` with c7aba79.

File: src/index.js [hunk 1]
```diff
@@ -82,7 +83,17 @@ function renderToString(vnode, context, opts, inner, isSvgMode) {
 				}
 			}
 
-			return renderToString(rendered, context, opts, opts.shallowHighOrder!==false);
```
Comment (developit) [inline src/index.js]: Idea: we can circumvent having to catch and re-throw the error when there's no componentDidCatch:

```js
if (c && c.componentDidCatch) {
	try {
		return renderToString(rendered, context, opts, opts.shallowHighOrder!==false);
	}
	catch (error) {
		c.componentDidCatch(error);
		rendered = c.render(c.props, c.state, c.context);
		// now we fall through to the non-componentDidCatch render!
	}
}
return renderToString(rendered, context, opts, opts.shallowHighOrder!==false);
```

**Comment by developit** [comment]
Just noticed the React docs seem to indicate they don't support componentDidCatch during SSR:
https://reactjs.org/docs/error-boundaries.html

Are we early on this and risking breaking compat?

**Comment by developit** [comment]
One thing I think we might want to consider is extending this to also allow catching enqueued state updates. Something like "render the component, then check if it errored or was dirtied via setState()/forceUpdate(). If so, render again".

**Comment by johnhaitas** [comment]
per react's server support for error boundaries (https://github.com/developit/preact-render-to-string/pull/66/#issuecomment-464513856) ... if react does not support it, then it probably is not right to support it

**Comment by marvinhagemeister** [comment]
I concur, this would be an awesome feature to support imo 👍

**Comment by rschristian** [comment]
Closing as implemented in #305, upon the work done here.

Thanks!

**Pull request closed without merging**